### DISCUSSION TITLE
Accessor/mutator doesn't change value

### DISCUSSION THREAD

**Discussion by francoism90** [livewire/livewire#5079] [7. Bugs (archived)]
```
\\ App\Models\Book.php
protected $appends = [
    'author',
];

protected function author(): Attribute
    {
        return Attribute::make(
            get: fn ($value) => $value ?? $this->author->getRouteKey(),
            set: fn ($value) => $value,
        );
    }

\\ LivewireEditBookComponent.php
protected function rules(): array
    {
        return [
            'book.author' => ['required'],
        ];
    }

<select wire:model="book.author">
@foreach ($this->authors as $author)
    <option value="{{ $author->getRouteKey() }}">{{ $author->name }}</option>
@endforeach
</select>
```
On init it has the correct binding, but on select it gets stuck on the first (pre)selected option.

Are accessors/mutators not supported? Are they readonly?

**Comment by michael-rubel** [comment]


**Reply by francoism90** [reply]
@michael-rubel Thanks for your detailed answer. However I'm a bit confused:

- Aren't you supposed to use [route binding](https://laravel-livewire.com/docs/2.x/rendering-components#route-model-binding)? Does this still expose the ID/Class/etc. from models?
- With `fill`/`loop` you purpose to simple map them directly instead, e.g. (`$this->fill(['name' => $post->name, ..])`), and use `updatedName`/`updatingName` to transform those values when needed, instead of using mutators?

Many thanks. :)

**Reply by michael-rubel** [reply]
@francoism90 ID/UUID will not be exposed, since you pass it during the Blade rendering stage. If you pass a whole model, it is additional work for Livewire internals.

The `fill` method will assign every attribute (with identical names) from the model to a property in the Livewire component (if defined). The loop functions are more flexible as it's able to work completely dynamically (without explicitly defining properties) and fetch data from nested arrays. But use it carefully and make sure you're not exposing sensitive data to the frontend. I once answered about the differences [here](https://github.com/michael-rubel/livewire-best-practices/discussions/2).

**Reply by francoism90** [reply]
@michael-rubel Thanks! I've inspected the requests, and it seems to fire with all parameters (id, name, etc.) if they are set as a rule. Is this correct? Is every 'public' variable/method accessible using `console.log`?

Could you please give an example how to use `fill`/`loop` when populating properties? `$this->fill($model)` would work, right?

**Reply by michael-rubel** [reply]
@francoism90 every public property defined in Livewire component is accessible from the browser.

Yes, `fill` or `propertiesFrom` should work.

**Reply by francoism90** [reply]
@michael-rubel Thanks for answering my questions, really appreciated. :)

Do you have any ideas how one could hide the real (model) id and let's say use a slug/uuid instead?

**Reply by michael-rubel** [reply]
@francoism90 If you don't want UUID to be accessible from the frontend at all, you can set it to protected/private property, but that way it would be unavailable in Blade.  If you want to keep it available in Blade, but inaccessible from the browser, you may use the [Livewire Access](https://github.com/archtechx/livewire-access) package.

**Reply by francoism90** [reply]
@michael-rubel Thanks! I'll checkout all your provided packages. :)
This would mean I still would be able to use ID's, but block access to them, right? Or do you recommend to always use UUID's on the front-end?

I wonder if more people having data leaks when using Livewire. On reading it does seem to offer some sort of protection against data manipulation, but I wasn't aware you could basically (console) log almost everything.

**Reply by michael-rubel** [reply]
@francoism90 Personally, I always use UUID for getting the model as it is considered more frontend-safe, etc. But I don't remove the default auto-incrementing ID, leaving it for internal DB purposes.

P.S. I'm looking to release a battle-tested code about auto-applying ordered UUID as a simple package. :)
